### PR TITLE
CPR-439 initial multiple record in cluster, reclustering

### PR DESCRIPTION
**Pull request by H-Iwanejko**
* Tests will be done as part of CPR-452

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/ReclusterService.kt [hunk 1]
```diff
@@ -5,39 +5,76 @@ import org.springframework.beans.factory.annotation.Value
...
     }
   }
 
+  private fun handleMultipleRecordsInCluster(personKeyEntity: PersonKeyEntity) {
+    checkClusterRecordsMatch(personKeyEntity)
+    // CPR-452
+  }
+
+  private fun checkClusterRecordsMatch(personKeyEntity: PersonKeyEntity): List<PersonEntity> {
+    val initialRecord = personKeyEntity.personEntities.first()
```
[the remark this answers is not in the document]
Comment (H-Iwanejko) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/ReclusterService.kt:49]: That is a good question. All the records in the cluster have to be checked for a link regardless. Whether we want to optimize which one we check first is something im not sure about.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/ReclusterService.kt [hunk 2]
```diff
@@ -5,39 +5,76 @@ import org.springframework.beans.factory.annotation.Value
...
+      initialMatchedRecords.isEmpty() -> recordsInClusterNotMatched.add(initialRecord)
+    }
+    return recordsInClusterNotMatched.toList()
+  }
+
+  private fun matchRecordAgainstCluster(recordToMatch: PersonEntity, personEntities: List<PersonEntity>): Pair<List<PersonEntity>, List<PersonEntity>> {
+    val recordsToMatch = personEntities.filterNot { it == recordToMatch }
+    val recordMatches = matchService.findHighConfidenceMatches(recordsToMatch, PersonSearchCriteria.from(recordToMatch))
+    val matchedRecords = recordMatches.map { it.candidateRecord }
+    val unmatchedRecords = recordsToMatch.filter { entity -> recordMatches.map { it.candidateRecord.id }.contains(entity.id).not() }
```
[the remark this answers is not in the document]
Comment (H-Iwanejko) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/ReclusterService.kt:74]: Yeah it does require some weird mapping coercion 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/ReclusterService.kt [hunk 3]
```diff
@@ -5,39 +5,81 @@ import org.springframework.beans.factory.annotation.Value
...
+
+    return recordsInClusterNotMatched.toList()
+  }
+
+  private fun matchRecordAgainstCluster(recordToMatch: PersonEntity, personEntities: List<PersonEntity>): Pair<List<PersonEntity>, List<PersonEntity>> {
+    val recordsToMatch = personEntities.filterNot { it == recordToMatch }
+    val matchedRecords = matchService.findHighConfidenceMatches(recordsToMatch, PersonSearchCriteria.from(recordToMatch)).map { it.candidateRecord }
+    val unmatchedRecords = recordsToMatch.filter { matchedRecord -> matchedRecords.map { it.id }.contains(matchedRecord.id).not() }
+    return Pair(matchedRecords, unmatchedRecords)
+  }
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/ReclusterService.kt:73]:
```suggestion
   private fun matchRecordAgainstCluster(recordToMatch: PersonEntity, personEntities: List<PersonEntity>): Pair<List<PersonEntity>, List<PersonEntity>> {
    val recordsToMatch = personEntities.filterNot { it == recordToMatch }
    return matchService.findHighConfidenceMatches(recordsToMatch, PersonSearchCriteria.from(recordToMatch))
      .map { it.candidateRecord }
      .partition { !recordsToMatch.contains(it) }
  }
```

Is this any better? The tests pass....

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/ReclusterService.kt [hunk 4]
```diff
@@ -5,39 +5,81 @@ import org.springframework.beans.factory.annotation.Value
...
+  private fun handleMultipleRecordsInCluster(personKeyEntity: PersonKeyEntity) {
+    checkClusterRecordsMatch(personKeyEntity)
+    // CPR-452
+  }
+
+  private fun checkClusterRecordsMatch(personKeyEntity: PersonKeyEntity): List<PersonEntity> {
+    val recordsInClusterMatched: MutableList<PersonEntity> = mutableListOf()
+    val recordsInClusterNotMatched: MutableList<PersonEntity> = personKeyEntity.personEntities.toMutableList()
+
+    personKeyEntity.personEntities.forEach { personEntity ->
```
[the remark this answers is not in the document]
Comment (H-Iwanejko) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/ReclusterService.kt:52]: Yeah it requires the loop as for example in a cluster A, B, C, D. Record A might only link to B and C. So then you would need to check if record D matches any. Say D matches to just C. We would keep it in the cluster 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/ReclusterService.kt (hunk 3 above)
Comment (H-Iwanejko) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/ReclusterService.kt]: Updated 👍 , need to be partitioned on the record match as the matchservice only returns high confidence matches.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/ReclusterService.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (markreesmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/ReclusterService.kt:52]: Fair enough - the code is quite complex but it sounds like the problem is also complex.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/ReclusterService.kt [hunk 5]
```diff
@@ -5,39 +5,80 @@ import org.springframework.beans.factory.annotation.Value
...
+      when {
+        recordsInClusterMatched.contains(personEntity) -> return@forEach
+      }
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/ReclusterService.kt:55]: I think I finally understand what this line does!

**Pull request merged**